mqtt-streaming/src/test/scala/docs/scaladsl/MqttSourceSpec.scala

```diff
@@ -0,0 +1,355 @@
+
+      publishFlow.complete()
+      // #at-least-once
+
+      streamCompletion.foreach { _ =>
+        mqttClientSession.shutdown()
+      }
+      // #at-least-once
+    }
+
+    "receive unacked messages later (when not using CleanSession)" in assertAllStagesStopped {
+      val testId = "5"
+      val time = LocalTime.now().toString
+      val clientId = s"streaming/source-spec/$testId/$time"
+      val topic = TopicPrefix + testId
+
+      val sessionSettings = MqttSessionSettings()
+      val input = Vector("one-" + time, "two-" + time, "three-" + time, "four-" + time, "five-" + time)
+      val ackedInFirstBatch = 2
+
+      val mqttClientSession: MqttClientSession = ActorMqttClientSession(sessionSettings)
+
+      // read first elements
+      val publishFlow: SourceQueueWithComplete[Command[Nothing]] = {
+        val subscriptions = MqttSubscriptions.atLeastOnce(topic)
+        val (switch, received) = MqttSource
+          .atLeastOnce(mqttClientSession,
+                       transportSettings,
+                       MqttRestartSettings(),
+                       MqttConnectionSettings(clientId).withConnectFlags(ConnectFlags.None),
+                       subscriptions)
+          .zipWithIndex
+          .mapAsync(1) {
+            case ((publish, ackHandle), index) if index < ackedInFirstBatch =>
+              ackHandle.ack().map(_ => publish)
+            case ((publish, _), _) =>
+              Future.successful(publish)
+          }
+          .viaMat(KillSwitches.single)(Keep.right)
+          .toMat(Sink.seq)(Keep.both)
+          .run()
+
+        val publishFlow = publish(topic, ControlPacketFlags.QoSAtLeastOnceDelivery, transportSettings, input)
+        sleepToReceiveAll()
+        switch.shutdown()
+        publishFlow.complete()
+        publishFlow
+      }
+      // read elements that where not acked
+      {
+        val subscriptions = MqttSubscriptions.atLeastOnce(topic)
+        val (switch, received) = MqttSource
+          .atLeastOnce(mqttClientSession,
+                       transportSettings,
+                       MqttRestartSettings(),
+                       MqttConnectionSettings(clientId).withConnectFlags(ConnectFlags.None),
+                       subscriptions)
+          .mapAsync(1) {
+            case (publish, ackHandle) =>
+              ackHandle.ack().map(_ => publish)
+          }
+          .viaMat(KillSwitches.single)(Keep.right)
+          .toMat(Sink.seq)(Keep.both)
+          .run()
+
+        sleepToReceiveAll()
+        switch.shutdown()
+
+        received.futureValue.map(_.payload.utf8String) should contain theSameElementsInOrderAs input.drop(
+          ackedInFirstBatch
+        )
+      }
+
+      publishFlow.watchCompletion().foreach { _ =>
+        mqttClientSession.shutdown()
+      }
+    }
+
+  }
+  private def sleepToReceiveAll(): Unit =
+    sleep(2.seconds, "to make sure we don't get more than expected")
```

Comment:
I've always found that sleeping in tests leads to flaky CI scenarios. Is there a better way?